README.md
# bcbio.variation.recall

Parallel merging, squaring off and ensemble calling for genomic variants.
Provide a general framework meant to combine multiple variant calls, either from
single individuals, batched family calls, or multiple approaches on the same
sample. Splits inputs based on shared genomic regions without variants, allowing
independent processing of smaller regions with variant calls. Handles:

- Merging multiple samples, called independently, into a single final VCF file.
- Squaring off multiple samples, called independently, by recalling at all
  identified genomic positions.
- Ensemble calling for samples called with multiple variant callers. Uses an
  intersection based approach, selecting variants present in at least `n` callers,
  where you choose `n`.

This is a work in progress.

## Usage

The executable `bcbio-variation-recall` bash script contains a ready to run jar
file. Pre-built distributions will be available. To create a development version
run `make` and the executable will be available in the `bin` directory. This
requires [leiningen].

[leiningen]: http://leiningen.org/

### Merge

    Merge multiple VCF files together, running in parallel over genomic regions.

    Usage: bcbio-variation-recall merge [options] out-file ref-file vcf-files

      out-file:  VCF (or bgzipped VCF) file to write merged output to
      ref-file:  FASTA format genome reference file
      vcf-files: VCF files to merge. Can be specified on the command line
                 or as a text file containing paths to files for processing

    Options:
      -c, --cores CORES    1  Number of cores to use
      -r, --region REGION     Genomic region to subset, in samtools format (chr1:100-200) or BED file
      -h, --help

### Squaring off

    Perform squaring off for a set of called VCF files, recalling at no-call positions in each sample.

    Usage: bcbio-variation-recall square [options] out-file ref-file [<vcf, bam, cram, or list files>]

      out-file:    VCF (or bgzipped VCF) file to write merged output to
      ref-file:    FASTA format genome reference file
      <remaining>: VCF files to recall and BAM or CRAM files for each sample. Can be specified
                   on the command line or as text files containing paths to files
                   for processing. VCFs can be single or multi-sample and BAM/CRAMs can be in
                   any order but each VCF sample must have an associated BAM/CRAM file to recall.
                   The sample names in the VCF file must match read groups in the BAM files.

    Options:
      -c, --cores CORES    1          Number of cores to use
      -m, --caller CALLER  freebayes  Calling method to use: samtools, freebayes, platypus
      -r, --region REGION             Genomic region to subset, in samtools format (chr1:100-200) or BED file
      -h, --help

### Ensemble

    Ensemble calling for samples: combine multiple VCF caller outputs into a single callset.

    Usage: bcbio-variation-recall ensemble [options] out-file ref-file [<vcf-files or list-files>]

       out-file:   bgzipped VCF file to write merged output to
       ref-file:   FASTA format genome reference file
      <remaining>: VCF files to include for building a final ensemble callset.
                   Specify on the command line or as text files containing paths to files.
                   VCFs can be single or multi-sample.
                   The input order of VCFs determines extraction preference in the final ensemble output.

    Options:
      -c, --cores CORES      1  Number of cores to use
      -n, --numpass NUMPASS  2  Number of callers a variant should be present in to pass
           --names NAMES        Comma separated list of names corresponding to VCFs, for annotating output
           --nofiltered         Remove filtered variants before performing ensemble calls
      -h, --help

## Thank you

External software provides the underlying algorithms. This tool is a framework
for pulling them together. The following command line programs need to be on
your path:

- [freebayes][freebayes]
- [vcflib][vcflib]
- [GATK MIT licensed framework][gatk-framework]
- [vt]
- [bedtools][bedtools]
- [bcftools 1.x][bcftools]
- [samtools 1.x][samtools]
- [sambamba]
- [platypus]

The [bcbio-nextgen][bcbio-nextgen] pipeline installs all this software automatically.

[bcbio-nextgen]: https://github.com/chapmanb/bcbio-nextgen
[bedtools]: http://bedtools.readthedocs.org/en/latest/
[vcflib]: https://github.com/ekg/vcflib
[bcftools]: https://github.com/samtools/bcftools
[freebayes]: https://github.com/ekg/freebayes
[gatk-framework]: https://github.com/chapmanb/gatk
[glia]: https://github.com/ekg/glia
[samtools]: http://samtools.sourceforge.net/
[platypus]: http://www.well.ox.ac.uk/platypus
[vt]: https://github.com/atks/vt
[scramble]: http://sourceforge.net/projects/staden/files/io_lib/
[sambamba]: https://github.com/lomereiter/sambamba

## License

The code is freely available under the [MIT license][l1].

[l1]: http://www.opensource.org/licenses/mit-license.html


HISTORY.md
## 0.2.6 (17 Sep 2019)

- Fix problem with not converting alleles to strings during v0.2.5 update.

## 0.2.5 (1 Sep 2019)

- Handle input VCFs with differing cases in REF and ALT alleles.

## 0.2.4 (22 Feb 2019)

- Fix platypus calling to correctly add contig headers to avoid GATK4 GatherVcf
  errors.

## 0.2.3 (12 Feb 2019)

- Fix wrapper script to avoid java in `/bin/java` without `JAVA_HOME` set.

## 0.2.2 (1 Feb 2019)

- Handle input to bcftools concat when all original files have no variants.
- Remove problematic INFO and FORMAT fields for FreeBayes. In complex decomposed, it
  retains more AD/AO/QA/GL fields then the headers allow, causing bcftools failures.
- Fix logging to correctly raise errors on failures.

## 0.2.1 (17 Jan 2019)

- Clarify filters on depth (DP) which fail on recent bcftools when present
  in both INFO and FORMAT key/values.

## 0.1.9 (3 August 2018)

- Remove usage of gatk-framework, which is no longer included in bcbio. Replace
  GATK usage with bcftools merge and concat.

## 0.1.8 (18 May 2018)

- Fix tabix index error for variants present at position 1 of contigs.

## 0.1.7 (15 August 2016)

- Avoid FreeBayes error when running `--variant-input` without an empty VCF.

## 0.1.6 (25 April 2016)

- Handle larger merge sizes with additional memory specifications up to 3000 or
  more samples.
- Use local temporary directories for GATK calls, avoiding filling up temporary
  space. Thanks to Roman Valls Guimerà.
- Correctly parse sample names from large numbers of inputs without causing
  overflow error.
- Do not silently exit if running into Java memory errors.
- Automatically convert ensemble output files to bgzipped output.

## 0.1.5 (15 April 2016)

- Add genotype qualities (GQ) to output of FreeBayes recalls.
- Use samtools instead of scramble for CRAM integration.

## 0.1.4 (15 October 2015)

- Annotate ensemble variant calls with names of callers supporting an ensemble variant.

## 0.1.3 (13 October 2015)

- Support latest GATK and htsjdk to handle issues with gVCFs on GATK 3.4 and
  different command line options in gatk-framework.

## 0.1.2 (25 April 2015)

- Correctly handle copies of single files in transactional directories.

## 0.1.1 (15 April 2015)

- Correctly intersect files when one of the inputs is empty. Thanks to Daryl
  Waggott.

## 0.1.0 (7 April 2015)

- Fix command line sorting issues with X/Y chromosomes. Thanks to Lorena
  Pantano.
- Separate command line help library into bcbio.run for use outside of
  bcbio.variation.recall.

## 0.0.9 (2 April 2015)

- FreeBayes: update with options for latest validated version 0.9.21-7

## 0.0.8 (20 March 2015)

- Avoid copy errors when merging recalls with only a single region.

## 0.0.7 (23 February 2015)

- Fix errors encountered on large runs, adding additional temporary directory
  usage during copy and spit commands to prevent partial files.

## 0.0.6 (30 January 2015)

- Correctly handle identically named input files for recalling.
- Avoid filtering low quality reference calls when recalling with Platypus.
- Correctly support vcffixup stdin in recent vcflib.
- Support new sambamba index changes.

## 0.0.5 (1 December 2014)

- Ensemble calling: handle variant inputs with multiple identical calls.

## 0.0.4 (28 October 2014)

- FreeBayes recalling fixes to support GATK compatibility: remove
  duplicate alternative alleles and ceil very low FreeBayes quality scores.

## 0.0.3 (26 October 2014)

- Fix FreeBayes recalling/squaring off to better distinguish reference and no
  call. Avoid filtering low-quality reference positions and assigning as no
  call while still removing problem variants.
- Fix ensemble calling for cases with closely spaced variants containing
  identical ref/alts.

## 0.0.2 (29 September 2014)

- Initial implementation of intersection based ensemble calling. Handle
  germline and tumor/normal cases.
- Use new picard and htsjdk libraries for variant manipulation.

## 0.0.1 (20 September 2014)

- Initial release with support for parallel genome merging and squaring
  off/joint calling with FreeBayes, Platypus and samtools.


src/bcbio/variation/recall/square.clj
(ns bcbio.variation.recall.square
  "Performing squaring off of variant call sets, recalling at all sample positions.
   This converts a merged dataset with no calls at positions not assessed in the
   sample, into a fully 'square' merged callset with reference calls at positions
   without evidence for a variant, distinguishing true no-calls from reference
   calls."
  (:require [bcbio.align.bam :as bam]
            [bcbio.align.cram :as cram]
            [bcbio.align.greads :as greads]
            [bcbio.run.clhelp :as clhelp]
            [bcbio.run.fsp :as fsp]
            [bcbio.run.itx :as itx]
            [bcbio.run.parallel :refer [rmap]]
            [bcbio.variation.ensemble.prep :as eprep]
            [bcbio.variation.recall.merge :as merge]
            [bcbio.variation.recall.vcfheader :as vcfheader]
            [bcbio.variation.recall.vcfutils :as vcfutils]
            [clojure.core.strint :refer [<<]]
            [clojure.java.io :as io]
            [clojure.java.shell :refer [sh]]
            [clojure.string :as string]
            [clojure.tools.cli :refer [parse-opts]]
            [me.raynes.fs :as fs]
            [version-clj.core :refer [version-compare]]))

(defn subset-sample-region
  "Subset the input file to the given region and sample."
  [vcf-file sample region out-file]
  (itx/run-cmd out-file
               "bcftools view -O ~{(vcfutils/bcftools-out-type out-file)} "
               "-r ~{(eprep/region->samstr region)} -s ~{sample} "
               "~{(eprep/bgzip-index-vcf vcf-file)} > ~{out-file}")
  (eprep/bgzip-index-vcf out-file :remove-orig? true))

(def ^{:doc "Filter to remove problematic annotations which can have incorrect numbers of fields
             in complex deletion/insertion cases with multiple alleles."}
  freebayes-field-filter
  "bcftools annotate -x FORMAT/AD,FORMAT/AO,FORMAT/QA,FORMAT/GL,INFO/QA | ")

(defn- intersect-variants
  "Retrieve VCF variants present in both in-file and cmp-file."
  [in-file cmp-file ref-file out-file caller]
  (let [field-filter (if (= :freebayes (keyword caller)) freebayes-field-filter "")]
    (cond
      (not (vcfutils/has-variants? in-file)) (itx/safe-copy in-file out-file)
      (not (vcfutils/has-variants? cmp-file)) (itx/safe-copy cmp-file out-file)
      :else (itx/run-cmd out-file
                         "vcfintersect -r ~{ref-file} -i ~{cmp-file} ~{in-file} | ~{field-filter}"
                         "bgzip > ~{out-file}")))
  (eprep/bgzip-index-vcf out-file :remove-orig? true))

(defn- unique-variants
  "Retrieve variants from in-file not present in cmp-file."
  [in-file cmp-file ref-file out-file]
  (cond
    (not (vcfutils/has-variants? in-file)) (itx/safe-copy in-file out-file)
    (not (vcfutils/has-variants? cmp-file)) (itx/safe-copy in-file out-file)
    :else (itx/run-cmd out-file
                       "vcfintersect -v -r ~{ref-file} -i ~{cmp-file} ~{in-file} | "
                       "bgzip > ~{out-file}"))
  (eprep/bgzip-index-vcf out-file :remove-orig? true))

(defmulti recall-variants
  "Recall variants only at positions in provided input VCF file, using multiple callers."
  (fn [& args]
    (keyword (get (last args) :caller :freebayes))))

(defmethod recall-variants :freebayes
  ^{:doc "Perform variant recalling at specified positions with FreeBayes.
          Cleans up FreeBayes calling:
           - Converts non-passing low quality variants to reference calls.
           - Reporting calls as no-call if they do not have at least a depth of 4
           - Ceil FreeBayes qualities at 1 to avoid errors when feeding to GATK downstream.
           - Remove duplicate alternative alleles."}
  [sample region vcf-file bam-file ref-file out-file config]
  (let [sample-file (str (fsp/file-root out-file) "-samples.txt")
        ploidy-str (if (:ploidy config) (format "-p %s" (:ploidy config)) "")
        filters ["NUMALT == 0", "%QUAL < 5", "AF[*] <= 0.5 && max(FORMAT/DP) < 4",
                 "AF[*] <= 0.5 && max(FORMAT/DP) < 13 && %QUAL < 10","AF[*] > 0.5 && max(FORMAT/DP) < 4 && %QUAL < 50"]
        nosupport-filter "bcftools filter -S . -e 'AC == 0 && max(FORMAT/DP) < 4' 2> /dev/null"
        filter_str (string/join " | " (map #(format "bcftools filter -S 0 -e 'AC > 0 && %s' 2> /dev/null" %)
                                           filters))]
    (spit sample-file sample)
    (if (vcfutils/has-variants? vcf-file)
      (itx/run-cmd out-file
                   "freebayes -b ~{bam-file} --variant-input ~{vcf-file} --only-use-input-alleles "
                   "--min-repeat-entropy 1 ~{ploidy-str} "
                   "--use-best-n-alleles 4 --min-mapping-quality 20 --genotype-qualities "
                   "-f ~{ref-file} -r ~{(eprep/region->freebayes region)} -s ~{sample-file}  | "
                   "~{freebayes-field-filter}"
                   "vcfuniqalleles | ~{filter_str} | vcffixup - | ~{nosupport-filter} | "
                   "awk -F$'\\t' -v OFS='\\t' '{if ($0 !~ /^#/ && $6 < 1) $6 = 1 } {print}' | "
                   "bgzip -c > ~{out-file}")
      (itx/safe-copy vcf-file out-file))
    (eprep/bgzip-index-vcf out-file)))

(defmethod recall-variants :platypus
  ^{:doc "Perform variant recalling at specified positions with Platypus.
          Perform post-filtration of platypus variant calls.
          Add a contig header line to output to satisfy GATK and bcftools downstream.
          Removes hard Q20 filter and replaces with NA12878/GiaB tuned depth
          and quality based filter.
          Performs normalization and removal of duplicate alleles.
          bgzips final output."}
  [sample region vcf-file bam-file ref-file out-file config]
  (let [filters ["FR[*] <= 0.5 && TC < 4 && %QUAL < 20",
                 "FR[*] <= 0.5 && TC < 13 && %QUAL < 10",
                 "FR[*] > 0.5 && TC < 4 && %QUAL < 20"]
        nosupport-filter "bcftools filter -S . -e 'TR == 0 && TC < 4' 2> /dev/null"
        filter_str (string/join " | " (map #(format "bcftools filter -S 0 -e 'TR > 0 && %s' 2> /dev/null" %) filters))]
    (itx/run-cmd out-file
                 "awk '{printf(\"##contig=<ID=%s,length=%d>\\n\",$1,$2);}' "
                 "~{ref-file}.fai > ~{out-file}-header.txt && "
                 "platypus callVariants --bamFiles=~{bam-file} --regions=~{(eprep/region->samstr region)} "
                 "--hapScoreThreshold 10 --scThreshold 0.99 --filteredReadsFrac 0.9 "
                 "--rmsmqThreshold 20 --qdThreshold 0 --abThreshold 0.0 --minVarFreq 0.0 "
                 "--refFile=~{ref-file} --source=~{vcf-file} --minPosterior=0 --getVariantsFromBAMs=0 "
                 "--logFileName /dev/null --verbosity=1 --output - | "
                 "awk -F$'\\t' -v OFS='\\t' '{if ($0 !~ /^#/) $7 = \"PASS\" } {print}' | "
                 "grep -v '##contig' | bcftools annotate -h ~{out-file}-header.txt | "
                 "vt normalize -r ~{ref-file} -q - 2> /dev/null | vcfuniqalleles | "
                 "~{filter_str} | vcffixup - | ~{nosupport-filter} | bgzip -c > ~{out-file}")
    (eprep/bgzip-index-vcf out-file)))

(defn vcf->bcftools-call-input
  "Convert a VCF input file into the custom format needed for bcftools call.
   XXX Not currently used as constrained calling is broken in samtools 1.0 for multiple positions."
  [vcf-file]
  (let [out-file (str (fsp/file-root vcf-file) "-callinput.tsv")]
    (itx/run-cmd out-file
                 "bcftools query -f '%CHROM\\t%POS\\t%REF,%ALT\\n' ~{vcf-file} > ~{out-file} ")
    (if false ;; (> (fs/size out-file) 0)
      (<< "-T ~{out-file} --constrain alleles")
      "")))

(defmethod recall-variants :samtools
  ^{:doc "Perform variant recalling at specified positions with samtools.
          XXX Curently uses normalization instead of constrained calling."}
  [sample region vcf-file bam-file ref-file out-file config]
  (let [filters ["AC[0] / AN <= 0.5 && max(FORMAT/DP) < 4 && %QUAL < 20"
                 "max(FORMAT/DP) < 13 && %QUAL < 10"
                 "AC[0] / AN > 0.5 && max(FORMAT/DP) < 4 && %QUAL < 50"]
        filter_str (string/join " | " (map #(format "bcftools filter -e '%s' 2> /dev/null" %) filters))]
    (itx/run-cmd out-file
                 "samtools mpileup -f ~{ref-file} -t DP -u -g -r ~{(eprep/region->samstr region)} "
                 "-l ~{vcf-file} ~{bam-file} | "
                 "bcftools call -m ~{(vcf->bcftools-call-input vcf-file)} - | "
                 "bcftools norm -f ~{ref-file} -m '+both' - | "
                 "sed 's/,Version=3>/>/' | sed 's/Number=R/Number=./' | "
                 "~{filter_str} | bgzip -c > ~{out-file}")
    (eprep/bgzip-index-vcf out-file)))

(defn union-variants
  ^{:doc "Merge single sample VCFs into combined VCF"}
  [vcf-files region ref-file out-file]
  (when (itx/needs-run? out-file)
    (let [input-list (str (fsp/file-root out-file) "-combineinputs.list")
          input-vcfs (filter vcfutils/has-variants? (map eprep/bgzip-index-vcf vcf-files))]
      (if (empty? input-vcfs)
        (itx/safe-copy (first vcf-files) out-file)
        (do
          (spit input-list (string/join "\n" input-vcfs))
          (itx/with-temp-dir [tmp-dir (fs/parent out-file)]
            (itx/run-cmd out-file
                         "bcftools concat --allow-overlaps "
                         "-O ~{(vcfutils/bcftools-out-type out-file)} "
                         "-r ~{(eprep/region->samstr region)} -f ~{input-list} "
                         "-o ~{out-file}"))))))
  (eprep/bgzip-index-vcf out-file))

(defn- sample-by-region
  "Square off a specific sample in a genomic region, given all possible variants.
    - Subset to the current variant region.
    - Identify missing uncalled variants: create files of existing and missing variants.
    - Recall at missing positions with FreeBayes.
    - Merge original and recalled variants."
  [sample vcf-file bam-file union-vcf region ref-file out-file config]
  (let [work-dir (fsp/safe-mkdir (str (fsp/file-root out-file) "-work"))
        fnames (into {} (map (fn [x] [(keyword x) (str (io/file work-dir (format "%s.vcf.gz" x)))])
                             ["region" "existing" "needcall" "recall"]))]
    (when (itx/needs-run? out-file)
      (itx/with-temp-dir [tmp-dir (fs/parent out-file)]
        (let [region-bam-file (greads/subset-in-region bam-file ref-file region tmp-dir)]
          (subset-sample-region vcf-file sample region (:region fnames))
          (intersect-variants (:region fnames) union-vcf ref-file (:existing fnames) (:caller config))
          (unique-variants union-vcf (:region fnames) ref-file (:needcall fnames))
          (recall-variants sample region (:needcall fnames) region-bam-file ref-file (:recall fnames) config)
          (union-variants [(:recall fnames) (:existing fnames)] region ref-file out-file))))
    out-file))

(defn- sample-by-region-prep
  "Prepare for squaring off a sample in a region, setup out file and check conditions.
   We only can perform squaring off with a BAM file for the sample."
  [sample vcf-file bam-file union-vcf region ref-file out-dir config]
  (let [out-file (str (io/file out-dir (format "%s-%s.vcf.gz" sample (eprep/region->safestr region))))]
    (cond
     (nil? bam-file) (subset-sample-region vcf-file sample region out-file)
     (itx/needs-run? out-file) (sample-by-region sample vcf-file bam-file union-vcf region ref-file out-file config)
     :else out-file)))

(defn- get-ploidy-line
  "Parse called ploidy from a VCF genotype call"
  [line]
  (let [[format call] (string/split line #"\t")
        gt-index (.indexOf (string/split format #":") "GT")]
    (when-not (neg? gt-index)
      (-> (string/split call #":")
          (nth gt-index)
          (string/split #"[|/]")
          count))))

(defn- get-ploidy-file
  "Retrieve ploidy from a single VCF file by grepping out FORMAT lines."
  [vcf-file region]
  (let [lines (string/split
               (:out (sh "bash" "-c"
                         (<< "bcftools view ~{vcf-file} -r ~{(eprep/region->samstr region)} "
                             "| grep -v '#' | cut -f 9-10 | head -5")))
               #"\n")]
    (remove nil? (map get-ploidy-line lines))))

(defn- get-existing-ploidy
  "Check ploidy in a region, cleanly handling recalling in mixed diploid/haploid (human chrM and sex chromosomes)"
  [vcf-files region]
  (apply max (mapcat #(get-ploidy-file % region) vcf-files)))

(defn by-region
  "Square off a genomic region, identifying variants from all samples and recalling at uncalled positions.
    - Identifies all called variants from all samples
    - For each sample, square off using `sample-by-region`
    - Merge all variant files in the region together."
  [vcf-files bam-files region ref-file dirs out-file config]
  (let [union-vcf (eprep/create-union :bcftools vcf-files ref-file region (:union dirs))
        config (assoc config :ploidy (get-existing-ploidy vcf-files region))
        region-square-dir (fsp/safe-mkdir (io/file (:square dirs) (get region :chrom "nochrom")
                                                   (eprep/region->safestr region)))
        region-merge-dir (fsp/safe-mkdir (str region-square-dir "-merge"))
        recall-vcfs (->> (rmap (fn [[sample vcf-file]]
                                 [sample (sample-by-region-prep sample vcf-file (get bam-files sample)
                                                                union-vcf region ref-file region-square-dir config)])
                               (mapcat (fn [vf]
                                         (for [s (vcfutils/get-samples vf)]
                                           [s vf]))
                                       vcf-files) (:cores config))
                         (into [])
                         (sort-by first)
                         (map second))]
    (merge/region-merge :bcftools recall-vcfs region ref-file out-file region-merge-dir)))

(defn- sample-to-bam-map*
  "Prepare a map of sample names to BAM files."
  [bam-files ref-file]
  (into {} (mapcat (fn [b]
                     (for [s (case (fs/extension b)
                               ".bam" (bam/sample-names b)
                               ".cram" (cram/sample-names b ref-file))]
                       [s b]))
                   bam-files)))

(defn sample-to-bam-map
  "Prepare a map of sample names to BAM files."
  [bam-files ref-file cache-dir]
  (let [cache-file (fsp/add-file-part (first bam-files) (format "%s-samplemap" (count bam-files))
                                      cache-dir ".edn")]
    (if (itx/needs-run? cache-file)
      (let [bam-map (sample-to-bam-map* bam-files ref-file)]
        (itx/with-tx-file [tx-cache-file cache-file]
          (spit tx-cache-file (pr-str bam-map)))
        (sample-to-bam-map bam-files ref-file cache-dir))
      (read-string (slurp cache-file)))))

(defn- check-versions
  "Ensure we have up to date versions of required software for recalling."
  [config]
  (when (= :freebayes (:caller config))
    (let [version (-> (sh "freebayes")
                      :out
                      (string/split #"\n")
                      last
                      (string/split #":")
                      last
                      string/trim)
          want-version "v0.9.14-1"]
      (when (neg? (version-compare version want-version))
        (throw (Exception. (format "Require at least freebayes %s for recalling. Found %s"
                                   want-version version)))))))

(defn combine-vcfs
  "Combine VCF files with squaring off by recalling at uncalled variant positions."
  [orig-vcf-files bam-files ref-file out-file config]
  (let [dirs {:union (fsp/safe-mkdir (io/file (fs/parent out-file) "union"))
              :square (fsp/safe-mkdir (io/file (fs/parent out-file) "square"))
              :inprep (fsp/safe-mkdir (io/file (fs/parent out-file) "inprep"))}
        bam-map (sample-to-bam-map bam-files ref-file (:inprep dirs))]
    (check-versions config)
    (merge/prep-by-region (fn [vcf-files region merge-dir]
                            (vcfutils/ensure-no-dup-samples vcf-files)
                            (by-region vcf-files bam-map region ref-file
                                       (assoc dirs :merge merge-dir) out-file config))
                          orig-vcf-files ref-file out-file config)))

(defn- usage [options-summary]
  (->> ["Perform squaring off for a set of called VCF files, recalling at no-call positions in each sample."
        ""
        "Usage: bcbio-variation-recall square [options] out-file ref-file [<vcf, bam, cram, or list files>]"
        ""
        "  out-file:    VCF (or bgzipped VCF) file to write merged output to"
        "  ref-file:    FASTA format genome reference file"
        "  <remaining>: VCF files to recall and BAM or CRAM files for each sample. Can be specified "
        "               on the command line or as text files containing paths to files "
        "               for processing. VCFs can be single or multi-sample and BAM/CRAMs can be in "
        "               any order but each VCF sample must have an associated BAM/CRAM file to recall."
        "               The sample names in the VCF file must match read groups in the BAM files."
        ""
        "Options:"
        options-summary]
       (string/join \newline)))

(defn -main [& args]
  (let [caller-opts #{:freebayes :platypus :samtools}
        {:keys [options arguments errors summary]}
        (parse-opts args [["-c" "--cores CORES" "Number of cores to use" :default 1
                           :parse-fn #(Integer/parseInt %)]
                          ["-m" "--caller CALLER" (str "Calling method to use: "
                                                       (string/join ", " (map name caller-opts)))
                           :default "freebayes"
                           :parse-fn #(keyword %)
                           :validate [#(contains? caller-opts %)
                                      (str "Supported calling options: "
                                           (string/join ", " (map name caller-opts)))]]
                          ["-r" "--region REGION"
                           "Genomic region to subset, in samtools format (chr1:100-200) or BED file"]
                          ["-h" "--help"]])]
    (cond
     (:help options) (clhelp/exit 0 (usage summary))
     errors (clhelp/exit 1 (clhelp/error-msg errors))
     (= 0 (count arguments)) (clhelp/exit 0 (usage summary))
     (< (count arguments) 3) (clhelp/exit 1 (usage summary))
     :else (let [[out-file ref-file & vcf-inputs] arguments
                 arg-files (clhelp/vcf-bam-args vcf-inputs)]
             (cond
              (not (empty? (:missing arg-files)))
              (clhelp/exit 1 (clhelp/error-msg (cons "Input files not found:" (:missing arg-files))))
              (or (not (fs/exists? ref-file)) (not (fs/file? ref-file)))
              (clhelp/exit 1 (clhelp/error-msg [(str "Reference file not found: " ref-file)]))
              :else
              (combine-vcfs (:vcf arg-files) (:bam arg-files) ref-file out-file options))))))
